galaxy/tools/deps/mulled/mulled_update_singularity_containers.py

```diff
@@ -0,0 +1,321 @@
+    # False
+    """
+    containers = []
+
+    repos_parameters = {'public': 'true', 'namespace': 'biocontainers'}
+    repos_headers ={'Accept-encoding': 'gzip', 'Accept': 'application/json'}
+    repos_response = requests.get(QUAY_API_ENDPOINT, headers=repos_headers, params=repos_parameters, timeout=12)
+
+    repos = repos_response.json()['repositories']
+    #repos = [n['name'] for n in repos]
+
+    for repo in repos:
+        logging.info(repo)
+        tags_response = requests.get("%s/biocontainers/%s" % (QUAY_API_ENDPOINT, repo['name']))
+        tags = tags_response.json()['tags']
+        for tag in tags:
+            containers.append('%s:%s' % (repo['name'], tag))
+
+    return containers
+
+def get_singularity_containers():
+    """
+    Gets all existing singularity containers from "https://depot.galaxyproject.org/singularity/"
+    # >>> lst = get_singularity_containers()
+    # >>> 'aragorn:1.2.36--1' in lst
+    # True
+    # >>> 'znc:latest' in lst
+    # False
+
+    """
+    index_url = "https://depot.galaxyproject.org/singularity/"
+    index = requests.get(index_url)
+    #l = response.text.split('\n')
+    tree = html.fromstring(index.content)
+    containers = tree.xpath('//a/@href')
+    containers = [container.replace('%3A', ':') for container in containers]
+    return containers
+
+def get_missing_containers(quay_list, singularity_list, blacklist_file=None):
+    """
+    Returns list of quay containers that do not exist as singularity containers. Files stored in a blacklist will be ignored
+    # >>> lst = get_missing_containers()
+    # >>> 'aragorn:1.2.36--1' in lst
+    # False
+    # >>> 'znc:latest' in lst
+    # False
+    # >>> 'pybigwig:0.1.11--py36_0' in lst
+    # True
+    # >>> 'samtools' in lst
+    # False
+    # >>> get_missing_containers(quay_list=[1, 2, 3, 'h', 'g', 'r'], singularity_list=[3, 4, 5], blacklist_file='blacklist.txt')
+    # [1, 2, 'h']
+
+    """
+    blacklist = []
+    if blacklist_file:
+        blacklist = open(blacklist_file).read().split('\n')
+    return [n for n in quay_list if n not in singularity_list and n not in blacklist]
+
+def docker_to_singularity(container):
+    """
+    Convert docker to singularity container
+    # >>> from glob import glob
+    # >>> glob('%s/abundancebin:1.0.1--0' % SINGULARITY_DESTINATION)
+    # []
+    # >>> docker_to_singularity('abundancebin:1.0.1--0')
+    # >>> glob('%s/abundancebin:1.0.1--0' % SINGULARITY_DESTINATION)
+    # ['summat/abundancebin:1.0.1--0']
+    """
+
+    try:
+        check_output("sudo %s build %s/%s docker://quay.io/biocontainers/%s && sudo rm -rf /root/.singularity/docker/" % (SINGULARITY_INSTALL, SINGULARITY_DESTINATION, container, container), stderr=subprocess.STDOUT, shell=True)
+    except subprocess.CalledProcessError as e:
+        error_info = {'code': e.returncode, 'cmd': e.cmd, 'out': e.output}
+        return error_info
+    else:
+        return None
+
+
+def get_test(container):
+    """
```

Comment:
Why do you use a get_test method here if this was done already in the extra file above?

Reply:
It's possible to do that. The other file does a much more thorough search for the tests - I'm not sure which is more appropriate. I thought the other one might be overkill.

Reply:
It's better to avoid code duplication, so you should at least merge both functions and depending on a parameter search deep or not. Code duplication in any way is bad.